### PR TITLE
add support to keep values of specific parts which is useful for example for metadata used in semver

### PR DESCRIPTION
**Pull request by martinm82**

### CODE REVIEW

File: .bumpversion.cfg [hunk 1]
```diff
@@ -1,15 +1,19 @@
 [bumpversion]
 commit = True
 tag = True
-current_version = 0.5.10
-parse = (?P<major>\d+)\.(?P<minor>\d+)\.(?P<patch>\d+)(\-(?P<release>[a-z]+))?
+current_version = 0.5.10-dev+2.3.4-meta
```
Comment (martinm82) [inline .bumpversion.cfg]: This was just for testing

File: .bumpversion.cfg [hunk 2]
```diff
@@ -1,15 +1,19 @@
...
+parse = (?P<major>\d+).(?P<minor>\d+).(?P<patch>\d+)(-(?P<release>[^\+]+))?(\+(?P<metadata>.*))?
 serialize = 
-	{major}.{minor}.{patch}-{release}
+	{major}.{minor}.{patch}-{release}+{metadata}
+	{major}.{minor}.{patch}+{metadata}
 	{major}.{minor}.{patch}
 
-[bumpversion:file:setup.py]
+[bumpversion:part:metadata]
+keep_value = True
```
Comment (florisla) [inline .bumpversion.cfg]: I would call this `independent`, because the `metadata` part can still change (if you change it explicitly).

File: .bumpversion.cfg [hunk 3]
```diff
@@ -1,15 +1,19 @@
...
 
-[bumpversion:file:bumpversion/__init__.py]
+#[bumpversion:file:setup.py]
```
Comment (florisla) [inline .bumpversion.cfg]: Beware, bump2version will remove these commented lines on the next bump!

**Comment by florisla** [comment]
Related issue: #80

File: .bumpversion.cfg (hunk 2 above)
Comment (c4urself) [reply .bumpversion.cfg]: hmm, I'm not sure if I would grok exactly what `independent` means at a first glance. Changing it explicitly is out-of-scope of bumpversion and in the context of bumpversion `keep_value: True` seems more straightforward.

**Comment by dalbani** [comment]
I've just discovered `bmp2version` when looking for a tool to  bump version of Helm charts.
And I have a question which may be related to this merge request: I would like to be able to set an (optional) static `-SNAPSHOT` suffix qualifier to a version. Kind of like Maven snapshot artifact are named.
For example:
- Starting point: `0.0.1-SNAPSHOT`.
- Then be able to release the chart as the developer sees fit, for example `0.0.1` or even `0.1.0`, etc.
- And finally afterwards the version would be set back to a (configurable) `-SNAPSHOT` version.

Basically kind of how the `jgitflow` plugin works, but not tied to a `pom.xml`.
Is there a way to accomplish that with `bump2version`? Would this `keep_value` functionality be needed then?
Thanks!

File: .bumpversion.cfg (hunk 2 above)
Comment (martinm82) [reply .bumpversion.cfg:14]: I have to agree with @c4urself as I as well wouldn't know how to interpret `independent`. But I am still open for other suggestions.

**Comment by martinm82** [comment]
Question is, is this implementation kind of acceptable or possibly causing some side-effects?

**Comment by martinm82** [comment]
> * Starting point: `0.0.1-SNAPSHOT`.
> * Then be able to release the chart as the developer sees fit, for example `0.0.1` or even `0.1.0`, etc.
Yes, that is possible with this tool.

> * And finally afterwards the version would be set back to a (configurable) `-SNAPSHOT` version.
You mean setting it back to `0.0.2-SNAPSHOT` or `0.1.1-SNAPSHOT`, right?

File: .bumpversion.cfg (hunk 2 above)
Comment (martinm82) [reply .bumpversion.cfg:14]: Ok, now I see where this `independent` actually comes from (ADVbumpversion). If you plan to merge both forks again I would be find with `independent`.
Are there any plans to do that?

**Comment by dalbani** [comment]
@martinm82 
Yes, that's what I meant.
Actually I think I've managed to have a working configuration (without this pull request):
```
[bumpversion]
parse = (?P<major>\d+)\.(?P<minor>\d+)\.(?P<patch>\d+)(\-(?P<release>[A-Z]+))?
serialize = 
	{major}.{minor}.{patch}-{release}
	{major}.{minor}.{patch}
search = version: {current_version}
replace = version: {new_version}
current_version = 0.1.0-SNAPSHOT

[bumpversion:part:release]
optional_value = PRODUCTION
values = 
	SNAPSHOT
	PRODUCTION
```
But, unless I'm doing it wrong, I had to add this unused `PRODUCTION` to make it work?!?
Otherwise I got such an error:
```
ValueError: The part has already the maximum value among ['SNAPSHOT'] and cannot be bumped.
```
It is what this pull request is (indirectly) going to fix?

**Comment by martinm82** [comment]
@dalbani yes this is the way to do that. It is basically the same as in all other implementations (eg. https://github.com/peritus/bumpversion).

BTW, for what you want to achieve you don't need the changes from this PR.

**Comment by dalbani** [comment]
@martinm82 
OK, good to know. But don't you find strange that this `PRODUCTION` entry has to be added, even though it's never used?!?

**Comment by ottomata** [comment]
@dalbani @martinm82 

I'm trying to accomplish the same thing but have not been successful.  I have 2 files with the version in it: .bumpversion.cfg and setup.cfg.
setup.cfg has e.g.
```
[metadata]
# ...
version = 0.1.0-dev
```

.bumpversion.cfg has
```
[bumpversion]
current_version = 0.1.0-dev
parse = (?P<major>\d+)\.(?P<minor>\d+)\.(?P<patch>\d+)(\-(?P<release>[A-Z]+))?
serialize =
    {major}.{minor}.{patch}-{release}
    {major}.{minor}.{patch}

[bumpversion:part:release]
optional_value = unused
values =
    dev
    unused

[bumpversion:file:setup.cfg]
```

If I `bump2version minor` I get versions serialized in both files as `0.2.0-dev` as expected.  If I do `bump2version release`, I get `ValueError: The part has already the maximum value among ['dev', 'unused'] and cannot be bumped.`.

From everything I've read, I'd expect `bump2version release` to change `0.1.0-dev` to `0.1.0`

Perhaps I need some way to set `first_value = <empty string>`? However I can't figure out how to to do that in the .ini file format.
